src/libstd/panicking.rs

```diff
@@ -28,11 +30,131 @@ thread_local! {
     }
 }
 
-fn log_panic(obj: &(Any+Send), file: &'static str, line: u32,
-             log_backtrace: bool) {
-    let msg = match obj.downcast_ref::<&'static str>() {
+static HANDLER_LOCK: StaticMutex = MUTEX_INIT;
```

Comment:
Perhaps this could use an rwlock? As written the `Sync` bound on the panic handler doesn't actually matter much (I think) as threads always have exclusive access tot he panic handler. I think that'd also remove the need for a check of only-lock-if-not-already-panicking.